### PR TITLE
Preserve user specified attributes and methods

### PR DESCRIPTION
**Pull request by bzinodev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38830 Preserve user specified attributes and methods**

This patch enables to preserve user specified attributes or non forward
methods. The API:
  _freeze_module(Module, ["a", "version"])

Differential Revision: [D21957316](https://our.internmc.facebook.com/intern/diff/D21957316)

### CODE REVIEW

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 1]
```diff
@@ -14,7 +14,33 @@ namespace {
...
+      for (auto& fn : module_.type()->methods()) {
+        if (fn->name() == name) {
+          preservedMethods_.insert(fn);
+          return true;
+        }
+      }
+      return false;
+    };
+
+    // forward is preserved by default.
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:36]: I don't know if we want this - if this is true there will be no way to remove the forward function. I think the signature should just be `freeze_module(module, List[str])` and then `freeze_module(module)` which will freeze all but forward.

File: torch/csrc/jit/passes/freeze_module.cpp [hunk 2]
```diff
@@ -36,17 +62,24 @@ class AttributePropagator {
...
     auto applyOptimizations = [](std::shared_ptr<Graph>& subgraph) {
       runOptimization(subgraph, /* unroll? */ false);
     };
-    optimizeSubGraphs(graph, applyInline);
-    propagateAttributes(graph);
-    optimizeSubGraphs(graph, applyOptimizations);
-    cleanupFrozenModule(graph);
+    for (auto function : preservedMethods_) {
+      auto graph = function->graph();
+      optimizeSubGraphs(graph, applyInline);
```
Comment (eellison) [inline torch/csrc/jit/passes/freeze_module.cpp:74]: This is sequentially runs `propagateAttributes` on each method, which records mutable attributes and then inlines the non-mutable ones. What we want instead is to first collect the set of all mutable attributes on all methods, and then inline them.

File: test/jit/test_freezing.py [hunk 3]
```diff
@@ -891,3 +891,47 @@ def test_freeze_module_detach_gradient(self):
...
+                return self.a + self.b
+
+        m = torch.jit.script(Module())
+        m.eval()
+        fm = torch._C._freeze_module(m._c, ["a"])
+        # Attribute "a" is preserved
+        self.assertTrue(fm.hasattr("a"))
+        self.assertFalse(fm.hasattr("b"))
+
+    def test_freeze_module_with_user_preserved_method(self):
```
Comment (eellison) [inline test/jit/test_freezing.py:912]: Can you add a test like:

        class Module(nn.Module):
            def __init__(self):
                super(Module, self).__init__()
                self.a = torch.tensor([1.1])
                self.b = torch.tensor([2.2])

            def forward(self, x):
                self.b[0] += 10
                return self.a + self.b

            @torch.jit.export
            def modify_a(self, x):
                self.a[0] += 10 
                return self. b + self.a

        m = torch.jit.script(Module())
        m.eval()
        fm = torch._C._freeze_module(m._c, ["modify_a"])
        FileCheck().check('prim::GetAttr[name="a"]').run(fm.forward.graph)
        FileCheck().check('prim::GetAttr[name="b"]').run(fm.modify_a.graph)

**Review by eellison — COMMENTED** [review]
Looks good - one question about the API and one question about the order we evaluate recordMutableAttrs in

File: torch/csrc/jit/python/init.cpp [hunk 4]
```diff
@@ -219,8 +219,11 @@ void initJITBindings(PyObject* module) {
       .def("_jit_pass_fold_convbn", &FoldConvBatchNorm2d)
       .def(
           "_freeze_module",
-          [](Module& module) { return freeze_module(module); },
-          py::arg("module"))
+          [](Module& module, std::vector<std::string> preservedAttrs) {
```
[the remark this answers is not in the document]
Comment (bzinodev) [reply torch/csrc/jit/python/init.cpp:229]: I think no. forward is expected to be always preserved

**Review by eellison — APPROVED** [review]
Code looks good to me, please update the one test I commented to be a little more confident it is working as expected

**Pull request closed without merging**